### ISSUE TITLE
error_index_generator in 1.24.0 fails to build using 1.24.0, entity not found

### ISSUE THREAD

**Issue by kyrias** [rust-lang/rust#48308]
I can build rust 1.24.0 using our 1.23.0 package, but when I try to build 1.24.0 again using the 1.24.0 package I already built I get the following error during `python2 ./x.py install`:

```
Documenting error index (x86_64-unknown-linux-gnu)
Building stage2 tool error_index_generator (x86_64-unknown-linux-gnu)
   Compiling error_index_generator v0.0.0 (file:///build/rust/src/rustc-1.24.0-src/src/tools/error_index_generator)
    Finished release [optimized] target(s) in 4.3 secs
thread 'main' panicked at 'entity not found', tools/error_index_generator/main.rs:209:9
note: Run with `RUST_BACKTRACE=1` for a backtrace.


command did not execute successfully: "/build/rust/src/rustc-1.24.0-src/build/x86_64-unknown-linux-gnu/stage2-tools-bin/error_index_generator" "html" "/build/rust/src/rustc-1.24.0-src/build/x86_64-unknown-linux-gnu/doc/error-index.html"
expected success, got: exit code: 101


failed to run: /build/rust/src/rustc-1.24.0-src/build/bootstrap/debug/bootstrap install
Build completed unsuccessfully in 0:01:47
```

**Comment by Mark-Simulacrum** [comment]
Looks like maybe we didn't generate the relevant metadata. Could you provide the full log by chance?

**Comment by kyrias** [comment]
@Mark-Simulacrum https://ptpb.pw/bkT0

**Comment by cuviper** [comment]
This was seen on Fedora too, in the build.log here:
https://koji.fedoraproject.org/koji/taskinfo?taskID=25212630

**Comment by cuviper** [comment]
To get away from distro idiosyncrasies, I tried default builds of stable, beta, and nightly, just pointing each to their matching `--local-rust-root ~/.rustup/toolchains/{stable,beta,nightly}...` for a local rebuild.  Stable (1.24) still failed, but beta and nightly were fine.

**Comment by cuviper** [comment]
The actual error via strace is:

```
open("tmp/extended-errors/x86_64-unknown-linux-gnu", O_RDONLY|O_NONBLOCK|O_DIRECTORY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
```

I strongly suspect we need 8c9bf663d4a8074e4955d1ad3f85a3b82e946297 from #47141.

**Comment by cuviper** [comment]
I backported that commit to stable, without the unrelated stage0-diagnostic stuff, and it seems to work:
https://github.com/cuviper/rust/commit/b445a52ea322758fb7b60fab5f890ef8c0f8df9c

**Comment by Mark-Simulacrum** [comment]
Thanks for experimenting! Since it's looking like we might be releasing 1.24.1 anyway (due to a variety of regressions that where detected post-release) I'll try to make sure that patch makes it into that.

**Comment by nikomatsakis** [comment]
triage: P-high

Gotta get this fixed.

**Comment by cuviper** [comment]
Fixed in 1.24.1.
